---
<h1 align = 'center'><font color = "#967BB6">Linear algebra and calculus in NumPy</font></h1>  

---

---
<h1 align = center> Importing Necessary Libraries </h1>

---

In [1]:
import numpy as np
import pandas as pd


---
<h1 align = center><font color = #5865F2 >Linear Algebra Tasks</font></h1>

---


---
<h1 align = center><font color =#D462FF > 1. Matrix Creation and Manipulation</font></h1>

---


---
<h2 align = center><font color =#CC338B >Create various types of matrices (zero matrix, identity matrix, random matrix).</font></h2>

---

---
<h2 align = center><font color = '#F62217'> Zero Matrix</font></h2>

---

#### What is Zero Matrix

A zero matrix is a matrix in which all the elements are zero. It is a special case of a square matrix.

In [2]:
# making a zero matrix of size 3x3
zeroMatrix = np.zeros(shape=(3,3), dtype='int32')
zeroMatrix

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]], dtype=int32)

---
<h2 align = center><font color = '#FF7900'> Identity Matrix</font></h2>

---

#### What is Identity Matrix ?

An identity matrix is a square matrix in which all the elements on the principal diagonal are ones (1s) and all other elements are zeros. It is also known as a unit matrix.

In [3]:
# making identity matrix of size 3x3

identityMatrix = np.eye(3, dtype='int32')
identityMatrix


array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]], dtype=int32)

---
<h2 align = center><font color = '#2672FF'> Random Matrix</font></h2>

---


#### What is Random Matrix ?

A random matrix is a matrix in which the elements are randomly generated numbers. It is often used in various applications, such as simulations, machine learning, and data analysis.

In [4]:
# making a random matrix of size 5x5

randomMatrix = np.random.randint(1, 25, size=(5,5), dtype='int32')
randomMatrix


array([[14,  7, 11, 12,  4],
       [22, 13,  2,  6, 20],
       [13, 21,  2, 17,  7],
       [ 6,  7,  1, 21, 23],
       [20, 10, 12, 16, 24]], dtype=int32)

---
<h2 align = center><font color =#CC338B >Perform basic matrix operations (addition, subtraction, multiplication).
</font></h2>

---


---
<h2 align = center><font color = '#6AFB92'>Addition</font></h2>

---

In [5]:
# Making 2 Random 5x5 Matrix

mat1 = np.random.randint(1,100 , size=(5,5),dtype='int16')
mat2 = np.random.randint(1,100 , size=(5,5),dtype='int16')

# Showing Mat1 and Mat2

print(f"\nMatrix 1\n")
print(mat1)

print(f"\nMatrix 2\n")
print(mat2)

# Adding the matrices

mat3 = mat1+mat2

print(f"\nResultant Matrix\n")
print(mat3)



Matrix 1

[[87 65 27 69 40]
 [ 8 99 52 56  5]
 [99 21 22 62 69]
 [93 67 44 93 72]
 [35 41 81  9 67]]

Matrix 2

[[46 51 20 31 80]
 [91 80 80 88  1]
 [62 43 27 17 22]
 [97  9 13 71 62]
 [63 37 45 22 70]]

Resultant Matrix

[[133 116  47 100 120]
 [ 99 179 132 144   6]
 [161  64  49  79  91]
 [190  76  57 164 134]
 [ 98  78 126  31 137]]


---
<h2 align = center><font color = '#FF7900'>Subtraction</font></h2>

---


In [6]:

# Making 2 Random 5x5 Matrix

mat1 = np.random.randint(1,100 , size=(5,5),dtype='int16')
mat2 = np.random.randint(1,100 , size=(5,5),dtype='int16')

# Showing Mat1 and Mat2

print(f"\nMatrix 1\n")
print(mat1)

print(f"\nMatrix 2\n")
print(mat2)

# Subtracting the matrices

mat3 = mat1-mat2

print(f"\nResultant Matrix\n")
print(mat3)



Matrix 1

[[94  6 19 64  7]
 [37 97  2 55 84]
 [16 77 71 81 28]
 [11 15  8 77 35]
 [72 84 25 63 82]]

Matrix 2

[[93 51 70 22 96]
 [84 60 99  7 58]
 [10 94 94 52 47]
 [28 78  2 12 47]
 [73 58 38 27 59]]

Resultant Matrix

[[  1 -45 -51  42 -89]
 [-47  37 -97  48  26]
 [  6 -17 -23  29 -19]
 [-17 -63   6  65 -12]
 [ -1  26 -13  36  23]]


---
<h2 align = center><font color = '#46C7C7'>Multiplication</font></h2>

---


In [7]:

# Making 2 Random 5x5 Matrix

mat1 = np.random.randint(1,100 , size=(5,5),dtype='int16')
mat2 = np.random.randint(1,100 , size=(5,5),dtype='int16')

# Showing Mat1 and Mat2

print(f"\nMatrix 1\n")
print(mat1)

print(f"\nMatrix 2\n")
print(mat2)

# Multiplying the matrices

mat3 = np.matmul(mat1,mat2)

print(f"\nResultant Matrix\n")
print(mat3)




Matrix 1

[[83  1 51 21 48]
 [91 10 28 16 16]
 [23 44 61 41 35]
 [45 95 11 58 63]
 [48 20 25 91 73]]

Matrix 2

[[54  1 13 31 91]
 [ 4 84 95 64 68]
 [ 3 41 82 37 78]
 [68 61 56 92 15]
 [89 69 22 55 93]]

Resultant Matrix

[[10339  6851  7588  9096 16378]
 [ 7550  4159  5677  6849 12873]
 [ 7504 11136 12547 11483 13713]
 [12394 16361 15146 16683 18142]
 [15432 13341 11276 16080 15832]]


---
<h2 align = center><font color =#CC338B >Transpose a matrix and find the determinant and inverse of a matrix.
</font></h2>

---



<h3 align = center> Making a random 10x10 Matrix </h3>




In [8]:
mat = np.random.randint(1,100 , size=(10,10),dtype='int16')

# Showing Matrix

print(f"\nMatrix 1\n")
print(mat)



Matrix 1

[[62 33 17 37 96 89 72 44 67 88]
 [66 67 24 43  8 65 64 43 48 10]
 [ 9 25 80 40 90  4 77 82 57 79]
 [56 81 97 83 20 56 35 49 71 73]
 [ 3 72  7 35 67 75 64  8 60 30]
 [55 69  6  9 19  3 51 81 95 99]
 [93  8 21 87 60  7  8 53 65 74]
 [46 87 21  3 81 88 72 55 20 32]
 [98 95 76 10 81 94 48 67 62 11]
 [ 3 63 25 78  9 29 63 14 39 89]]


---

<h2 align = center> <font color = '#38ACEC'> Transposing the matrix </font></h2>

---


#### What is Transposing Matrix ?

A transposed matrix is obtained by swapping the rows and columns of the original matrix. 


In [9]:
transposedMatrix = np.transpose(mat)
print(f"\nTransposed Matrix\n")
print(transposedMatrix)


Transposed Matrix

[[62 66  9 56  3 55 93 46 98  3]
 [33 67 25 81 72 69  8 87 95 63]
 [17 24 80 97  7  6 21 21 76 25]
 [37 43 40 83 35  9 87  3 10 78]
 [96  8 90 20 67 19 60 81 81  9]
 [89 65  4 56 75  3  7 88 94 29]
 [72 64 77 35 64 51  8 72 48 63]
 [44 43 82 49  8 81 53 55 67 14]
 [67 48 57 71 60 95 65 20 62 39]
 [88 10 79 73 30 99 74 32 11 89]]


---
<h2 align = center> <font color = '#00A36C'> Finding Determinant </font></h2>

---

#### What is Determinant?

The determinant of a square matrix is a scalar value that can be computed from the elements of the matrix and has many important properties. 

In [10]:
determinant = np.linalg.det(mat)
print(f"\nDeterminant of Matrix 1: {determinant}\n")


Determinant of Matrix 1: 6.74143640098782e+18



---
<h2 align = center> <font color = '#FF6700'> Matrix Inverse </font></h2>   

---
#### What is Matrix Inverse?

The inverse of a square matrix is another square matrix that has the same dimension as the original matrix and satisfies the following property:

mat * inverseMat = inverseMat * mat = I

where I is the identity matrix of the same size as the original matrix.

In [11]:
inverseMatrix = np.linalg.inv(mat)
print(f"\nInverse of Matrix\n")
print(inverseMatrix)



Inverse of Matrix

[[ 2.98485304e-03 -6.56494185e-04 -3.26160883e-03 -1.99330647e-02
  -1.22528894e-02 -1.57869810e-03  2.14327502e-03 -1.41072563e-02
   2.56115872e-02  2.23781897e-02]
 [-1.19405850e-02 -7.79766351e-03 -3.83419596e-03 -4.76850492e-03
   3.49563618e-03  2.80926290e-03  2.73435328e-03  4.36730041e-03
   9.22642310e-03  1.07098644e-02]
 [ 2.65777316e-03 -3.71184217e-03  3.85018763e-03 -1.60619049e-03
  -6.57875527e-03 -3.60608842e-03 -6.06293860e-03 -1.12916517e-02
   1.43786725e-02  9.24170551e-03]
 [-6.45078272e-03  6.45442335e-03  1.04465457e-03  8.31596222e-03
   6.55112963e-03 -4.68088105e-03  8.76195383e-03  8.67276541e-03
  -1.48383123e-02 -7.66614173e-03]
 [-3.56193703e-03 -8.97417388e-03  2.24073889e-03 -8.49383525e-03
   5.00487829e-03 -2.78757918e-03  5.86718483e-03  7.89313117e-05
   8.59989895e-03  4.95226316e-03]
 [ 9.95994500e-03  5.36383662e-03 -4.17337505e-03  2.33112687e-02
   3.35204587e-03 -9.98794088e-04 -4.28248825e-03  1.40857995e-02
  -2.41938847

---
<h1 align = center><font color =#D462FF > 2. Solving Linear Equations</font></h1>

---



---
<h2 align = center> <font color = '#FF0080'> Use NumPy to solve a system of linear equations. </font></h2>

---

#### What is Linear Equation?

A linear equation is an equation that can be written in the form of ax + by = c, where a, b, and c are constants and x and y are variables.





In [12]:
# Example Equation :
# x – y = 12 ---(1) 
# 2x + y = 22 ---(2)

# Converting the equation into matrix form Ax = b

a = np.array([[1,-1],[2,1]])
b = np.array([12,22])

x = np.linalg.solve(a,b)
print(x)


[11.33333333 -0.66666667]


#### Picture Solution

<img src="sol.png" alt="Linear Equations">




### Another Linear Equation Solution

In [13]:
# Equation
#  2x + 4y = 5
#  6x + 8y = 6

a = np.array([[2,4],[6,8]])
b= np.array([5,6])

x = np.linalg.solve(a,b)
x


array([-2.  ,  2.25])

---

<h2 align = center> <font color = '#0A7A8A'> Implement matrix factorization methods (LU decomposition, QR decomposition). </font></h2>

---
#### What is Matrix Factorization?

Matrix factorization is a method for breaking down a matrix into simpler matrices. There are two main types of matrix factorization:

1. LU Decomposition: This method decomposes a square matrix into a lower triangular matrix (L) and an upper triangular matrix (U).

2. QR Decomposition: This method decomposes a square matrix into a orthogonal matrix (Q) and a upper triangular matrix (R).

#### What is lower triangular matrix ?

A lower triangular matrix is a square matrix in which all elements above the main diagonal are zero.

#### What is upper triangular matrix ?

An upper triangular matrix is a square matrix in which all elements below the main diagonal are zero.



In [14]:
### Making a random square matrix

mat = np.random.randint(1,100 , size=(3,3),dtype='int16')

# Showing Matrix

print(f"\nMatrix 1\n")
print(mat)



Matrix 1

[[99 27 25]
 [51 99  3]
 [47 65 35]]


---
<h2 align = center> <font color = '#38ACEC'> LU Decomposition </font></h2>

---
#### Applications:

LU decomposition is used in various applications such as solving systems of linear equations, computing the inverse of a matrix, and solving linear least squares problems. It is particularly useful for solving systems of linear equations with a large number of variables.





In [15]:
import scipy.linalg
permutation,lower,upper = scipy.linalg.lu(mat)

print(f"\nLower Triangular Matrix (L):\n")
print(lower)
print(f"\nUpper Triangular Matrix (U):\n")
print(upper)

# As we know we can get our orignal matrix back by multiplying permutation matrix with lower triangular matrix .
# Then multiplying the resultant with upper triangular matrix. 

Q=np.dot(permutation,lower)

print("\nFinal Resultant Matrix\n")
print(np.dot(Q,upper))




Lower Triangular Matrix (L):

[[1.         0.         0.        ]
 [0.5151515  1.         0.        ]
 [0.47474748 0.6132478  1.        ]]

Upper Triangular Matrix (U):

[[99.       27.       25.      ]
 [ 0.       85.09091  -9.878788]
 [ 0.        0.       29.189459]]

Final Resultant Matrix

[[99. 27. 25.]
 [51. 99.  3.]
 [47. 65. 35.]]


---

<h2 align = center> <font color = '#00A36C'> QR Decomposition </font></h2>

---

#### Usage:

QR decomposition is used in various applications such as solving systems of linear equations, computing the singular value decomposition (SVD), and finding the nearest matrix to a given matrix. It is particularly useful for solving linear least squares problems and computing the eigenvalues and eigenvectors of a matrix.



In [16]:
mat = np.random.randint(1,1000 , size=(4,4), dtype='int16')

# Showing Matrix

print(f"\nMatrix 1\n")
print(mat)

# QR Decomposition

Q,R = np.linalg.qr(mat)

print(f"\nOrthogonal Matrix (Q):\n")
print(Q)

print(f"\nUpper Triangular Matrix (R):\n")
print(R)

# Multiplying Q and R to get original matrix

print("\nFinal Resultant Matrix\n")
print(np.dot(Q,R))







Matrix 1

[[ 37  68 599  73]
 [909 890 989 509]
 [211 257  33 987]
 [733 357 468 630]]

Orthogonal Matrix (Q):

[[-0.03116559 -0.12917817  0.91235484 -0.38723424]
 [-0.76566273 -0.54790214  0.03836995  0.33480081]
 [-0.17772809 -0.29674641 -0.40692038 -0.84544077]
 [-0.6174156   0.77139991  0.02349902 -0.15227594]]

Upper Triangular Matrix (R):

[[-1187.20680591  -949.65257476 -1070.72415157  -956.38686903]
 [    0.          -297.2910817   -268.03041702   -95.21895914]
 [    0.             0.           582.01760007  -300.69382133]
 [    0.             0.             0.          -788.23837281]]

Final Resultant Matrix

[[ 37.  68. 599.  73.]
 [909. 890. 989. 509.]
 [211. 257.  33. 987.]
 [733. 357. 468. 630.]]


<h2 align = center> <font color = 'gold'> Application </font></h2>

#### What is singular matrix ? 

A singular matrix is a square matrix that does not have an inverse. In other words, its determinant is zero. This means that the system of linear equations Ax = b has either no solution, infinitely many solutions, or a unique solution depending on the values of the constants a1, a2,..., an.


In [17]:

# For example, we can use QR decomposition to solve a system of linear equations.

# Equation

#  2x + 9y + 6z = 10
#  3x + 6y + 9z = 15
#  4x + 8y + 12z = 20

a = np.array([[2,9,6],[3,6,9],[4,8,12]])
b= np.array([10,15,20])

# sol = np.linalg.solve(a,b) This Can't be run because the matrix is singular. QR decomposition or other methods can be used to find a least-squares solution in such cases.

# Perform QR decomposition
Q, R = np.linalg.qr(a)

# Solve for x, y, z
QT_b = np.dot(Q.T, b)
x = np.linalg.solve(R, QT_b)

print(f"\nSolution: x = {x[0]}, y = {x[1]}, z = {x[2]}\n")




Solution: x = 19.99999999999999, y = -2.869792300935271e-15, z = -4.999999999999996



---
<h1 align = center><font color =#AABB44 > 3. Eigenvalues and Eigenvectors</font></h1>

---



#### What is Eigenvalues ? 

Eigenvalues are scalar values associated with a square matrix. They are solutions to the characteristic equation det(A - λI) = 0, where A is the matrix and λ is the eigenvalue. Eigenvalues can be found using the numpy.linalg.eig() function.

#### What is Eigenvectors ?

Eigenvectors are non-zero vectors that, when multiplied by a square matrix, produce a scaled version of themselves. They are related to eigenvalues and are found using the numpy.linalg.eig() function.

#### What is the relationship between Eigenvalues and Eigenvectors ?

Eigenvalues and eigenvectors have a significant relationship in linear algebra. They are used to analyze the properties of linear transformations and solve systems of linear equations. For example, eigenvalues and eigenvectors can be used to determine the stability of a system of linear differential equations, to find the principal components of a dataset, and to perform PCA (Principal Component Analysis) for dimensionality reduction.



---

<h2 align = center> <font color = 'Orchid'> Calculate the eigenvalues and eigenvectors of a given matrix.. </font></h2>

---

In [18]:
mat = np.array([[2, 1, 0],
                [1, 3, 1],
                [0, 1, 2]])

# Calculate eigenvalues and eigenvectors

eigenvalues, eigenvectors = np.linalg.eig(mat)

print(f"\nEigenvalues:\n")
print(eigenvalues)

print(f"\nEigenvectors:\n")

for i in range(len(eigenvalues)):
    print(f"\nEigenvector {i+1}:\n")
    print(eigenvectors[:,i])


Eigenvalues:

[4. 2. 1.]

Eigenvectors:


Eigenvector 1:

[-0.40824829 -0.81649658 -0.40824829]

Eigenvector 2:

[ 7.07106781e-01 -3.45742585e-16 -7.07106781e-01]

Eigenvector 3:

[ 0.57735027 -0.57735027  0.57735027]


---

<h2 align = center> <font color = 'Lime'> Verify the results by reconstructing the original matrix.
 </font></h2>

 ---

In [19]:
# Formula for reconstructing the original matrix

#A = P Λ P^-1.

# Reconstruct the original matrix

A = np.diag(eigenvalues)
P = eigenvectors

reconstructedMatrix = np.dot(np.dot(P,A),np.linalg.inv(P))


print(f"\nReconstructed Matrix:\n")
print(reconstructedMatrix)



Reconstructed Matrix:

[[2.00000000e+00 1.00000000e+00 3.34106299e-16]
 [1.00000000e+00 3.00000000e+00 1.00000000e+00]
 [2.27686080e-16 1.00000000e+00 2.00000000e+00]]


---
<h1 align = center><font color =#BF4311 > 4. Vector Operations</font></h1>

---



#### What is a Vector ? 

A vector is a geometric object that has magnitude (length) and direction. In other words, a vector is a one-dimensional array of numbers. Vectors can be represented using arrow notation, where the arrow starts at the origin (0, 0, 0) and points in the specified direction.

---

<h2 align = center> <font color = 'purple'> Perform basic vector operations (addition, dot product, cross product). </font></h2>

---



### Addition

In [20]:
# Making 2 vectors

v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])

# Adding vectors

result_addition = v1 + v2

print(f"\nAddition Result:\n")
print(result_addition)



Addition Result:

[5 7 9]


### Dot Product

what is dot product ? 

dot product is a scalar value that represents the magnitude of one vector in the direction of another vector. It is calculated by multiplying the corresponding components of the two vectors and then summing the results.

Example Code is given below:

In [21]:
# Dot product

result_dot_product = np.dot(v1, v2)

print(f"\nDot Product Result:\n")
print(result_dot_product)


Dot Product Result:

32


### Cross Product

what is cross product? 

cross product is a vector that results from the multiplication of two vectors. It is calculated by finding the determinant of a 2x2 matrix formed by the two vectors.

Example Code is given below:

In [22]:
#Cross product

v1 = np.array([1, 2, 3])
v2 = np.array([4, 5, 6])



result_cross_product = np.cross(v1, v2)

print(f"\nCross Product Result:\n")
print(result_cross_product)



Cross Product Result:

[-3  6 -3]


---

<h2 align = center> <font color = 'MintCream'> Normalize a vector and compute vector norms. </font></h2>

---

#### What is normalization of a vector?

Normalization is the process of scaling a vector to have a magnitude of 1. It is done by dividing the vector by its length.

#### What is vector norms ?

Vector norms are measures of the size or length of a vector. There are several types of vector norms, such as:

- Euclidean norm (L2 norm): sqrt(sum of squares of vector components)
- Manhattan norm (L1 norm): sum of absolute values of vector components
- Maximum norm (L Infinity norm): maximum absolute value of vector components

---
#### Vector Normalization
---

In [23]:
# Making a vector

v = np.array([8, 22, 100])

# Normalizing the vector

normalized_v = v / np.linalg.norm(v)

print(f"\nNormalized Vector:\n")
print(normalized_v)



Normalized Vector:

[0.07789417 0.21420895 0.97367706]


#### Results

As we can see, the normalized vector has a magnitude of 1, and its elements are scaled such that the vector points in the same direction as the original vector.

---
#### Using Euclidean norm

---

In [24]:

# Calculating Euclidean norm

euclideanNorm = np.linalg.norm(v)

print(f"\nEuclidean Norm:\n")
print(euclideanNorm)

normalized_v = v / euclideanNorm

print(f"\nNormalized Vector Using Euclidean Norm:\n")
print(normalized_v)


Euclidean Norm:

102.70345661174214

Normalized Vector Using Euclidean Norm:

[0.07789417 0.21420895 0.97367706]


---
### Using Manhattan Norm

---

In [25]:


manhattanNorm = np.linalg.norm(v, ord=1)

print(f"\nManhattan Norm:\n")
print(manhattanNorm)
# Using Maximum Norm

normalized_v = v / manhattanNorm

print(f"\nNormalized Vector Using Manhattan Norm:\n")
print(normalized_v)




Manhattan Norm:

130.0

Normalized Vector Using Manhattan Norm:

[0.06153846 0.16923077 0.76923077]


In [26]:
# Using Infinity Norm

infinityNorm = np.linalg.norm(v, ord=np.inf)

print(f"\nMaximum Norm:\n")
print(infinityNorm)

normalized_v = v / infinityNorm

print(f"\nNormalized Vector Using Infinity Norm:\n")
print(normalized_v)




Maximum Norm:

100.0

Normalized Vector Using Infinity Norm:

[0.08 0.22 1.  ]


---
<h1 align = center><font color =#5865F2 > 5. Matrix Decomposition</font></h1>

---






---
<h2 align = center> <font color = 'blurple'> Understand and implement Principal Component Analysis (PCA) using SVD.
  </font></h2>

---




#### What is Principal Component Analysis ?

Principal Component Analysis (PCA) is a statistical technique used to reduce the dimensionality of a dataset while retaining as much information as possible. It finds the principal components, or directions of maximum variance in the data, and projects the data onto these principal components to create a new, lower-dimensional representation.

#### How does PCA work?

1. Center the data: Subtract the mean of each feature from the corresponding data points to make the mean of each feature zero.

2. Compute the covariance matrix: Calculate the covariance matrix of the centered data, which represents the correlation between the features.

3. Compute the singular value decomposition (SVD): Perform the singular value decomposition (SVD) of the covariance matrix, using the numpy.linalg.svd() function. The SVD factors the covariance matrix into U, S, and V^T, where U and V^T are orthogonal matrices and S is a diagonal matrix with non-negative real numbers on the diagonal.

4. Select the desired number of principal components: Select the desired number of principal components based on the desired explained variance ratio. The explained variance ratio is the proportion of the total variance that each principal component captures.

5. Project the data onto the principal components: Multiply the centered data by the transpose of the selected principal components matrix (V^T) to obtain the new, lower-dimensional representation.



#### What are benefits of Principal Component Analysis?

- Reduces dimensionality: PCA helps reduce the dimensionality of a dataset by identifying the principal components that capture the most variance in the data.

- Enhances data interpretability: Principal components are orthogonal, meaning they are uncorrelated, which makes them easier to interpret and visualize.

- Improves computational efficiency: PCA can be computationally more efficient than other dimensionality reduction techniques, such as feature selection.

- Preserves important information: PCA retains as much information as possible from the original dataset during the dimensionality reduction process, ensuring that the reduced representation retains the essential patterns and relationships.



### What is covariance? 

Covariance is a measure of the linear relationship between two random variables. It quantifies the degree to which the two variables change together. This is very similar to correlation but there is no bounding of range(-1 to 1).

### What is covariance matrix?

Covariance matrix is a square matrix that represents the covariance between each pair of features in a dataset.



#### What is Singular Value Decomposition (SVD) ?

Singular Value Decomposition (SVD) is a matrix decomposition technique that factorizes a matrix into three matrices: U, Singular, and V^T. The U matrix contains the left singular vectors, the Singular matrix contains the singular values, and the V^T matrix contains the right singular vectors.

## Implementation

### Loading Dataset 

In [27]:
# Generating random data
n_samples = 100
n_features = 50
randomData = np.random.randint(1,1000,size=(n_samples, n_features),)

data = pd.DataFrame(randomData, columns=[f'Feature_{i+1}' for i in range(n_features)])
print(data.head())



   Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  Feature_6  \
0         55        565        803        156        993         33   
1        641         13        608        779        178        334   
2        189        106        671        375        488        770   
3        863        471        131        283        829        884   
4        795        367        855        889        929        153   

   Feature_7  Feature_8  Feature_9  Feature_10  ...  Feature_41  Feature_42  \
0        910        962        309         686  ...         509         493   
1        961        437         78         787  ...         295         864   
2        646        549        954         568  ...         825         874   
3        850        805        606         513  ...         356         875   
4        366        881        574         420  ...         870         575   

   Feature_43  Feature_44  Feature_45  Feature_46  Feature_47  Feature_48  \
0          38        

### Making Random Target Variable

In [28]:
target = np.random.randint(0, 2, size=n_samples)


### Training Data

In [29]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier


# Splitting the data into training and testing sets


X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=42)


### Training Linear Regression Model

model = KNeighborsClassifier()

model.fit(X_train, y_train)

# Predicting on the test data

model.score(X_test,y_test)







0.35

### Normalizing Data

In [30]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

normalizedData = scaler.fit_transform(data)



### Finding Covariance Matrix

In [31]:
# Covalence can be found using transpose of normalizedData
covarianceMatrix = np.cov(normalizedData.T)

## Using SVD 

In [35]:


from numpy.linalg import svd

U, S, VT = svd(covarianceMatrix)


eigenValues = S**2
eigenVectors = VT.T

## Here we are finding ratio variance that a singular value explains
explainedVariance = S / np.sum(S)

## Here we are finding cumulative sum of explained variance.
cumulativeVariance = np.cumsum(explainedVariance )



k = np.argmax(cumulativeVariance >= 0.95) + 1



principalComponents = VT.T[:,:k]

transformedData = np.dot(normalizedData, principalComponents)

transformedData.shape

transformed_df = pd.DataFrame(transformedData, columns=[f'PC{i+1}' for i in range(k)])

print(transformed_df.shape)



# Splitting the data into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(transformed_df, target, test_size=0.2, random_state=42)

# Training the model

model.fit(X_train, y_train)

# Predicting on the test data

model.score(X_test, y_test)





[0.056521   0.10432247 0.15071886 0.19368208 0.23579358 0.27594389
 0.31501684 0.35177655 0.38780377 0.42242575 0.45627618 0.4887619
 0.52042263 0.54866054 0.5764473  0.6029436  0.62816531 0.65206618
 0.67522888 0.69706514 0.71818377 0.738138   0.75738341 0.77546554
 0.79252319 0.80769864 0.82236769 0.83648994 0.85013001 0.86338759
 0.87616591 0.88798553 0.89902291 0.90931199 0.91897972 0.92850288
 0.93701379 0.94536179 0.9531522  0.9603325  0.9665607  0.97210224
 0.97723433 0.98166925 0.98597041 0.98934857 0.99237785 0.99528899
 0.99774675 1.        ]
(100, 39)


0.3

## Interpretation

The Principal Component Analysis (PCA) has been successfully applied to the dataset. The covariance matrix has been computed, and Singular Value Decomposition (SVD) has been performed to obtain the principal components. The transformed data has been obtained by projecting the original data onto the principal components.

The Model with all the features and The model with 40 features giving the same score. 

The final transformed dataset has 100 samples and 40 principal components. The model has been trained using the transformed data and evaluated on the test set. The accuracy of the model is approximately 50% , indicating that the PCA dimensionality reduction technique has successfully preserved the essential patterns and relationships in the original dataset.




---
<h1 align = center><font color = #64E986 >Calculus Tasks</font></h1>

---



---
<h2 align = center> <font color = '#FE632A'> 

 1.Numerical Differentiation

  </font></h2>

---


---

<h2 align = center> <font color = 'pink'> 

Use NumPy to compute the numerical derivative of a given function.
  </font></h2>

  
---


#### What is Numerical derivative of function?

Numerical derivative is a technique used to approximate the derivative of a function by computing the difference between the function values at two points, divided by the difference between the two points.



In [122]:
# Given function

# f(x) = x^3 + x^2 + 5

def f(x):
    return x**3 + x**2 + 5



# Numerical Derivative 

def numericalDerivative(f, x, h=1e-6):
    return (f(x + h) - f(x - h)) / (2 * h)

# Compute numerical derivative using forward difference method

x = np.arange(1,5)


# Compute numerical derivative using central difference method

print(f"\nNumerical derivative using central difference method at x = {x}:")
print(numericalDerivative(f, x))



# Now Checking the derivative using analytical method
def derivative(x): # Returning the derivative of function
    return 3 * x**2 + 2 * x 

print(f"\nAnalytical derivative at x = {x}:")
print(derivative(x))


# As we can see analytical derivatives and numerical derivatives are approximately same






Numerical derivative using central difference method at x = [1 2 3 4]:
[ 5.         16.         33.00000001 56.00000001]

Analytical derivative at x = [1 2 3 4]:
[ 5 16 33 56]


<h2 align = center> <font color = 'applegreen'> 

Implement forward, backward, and central difference methods for differentiation

  </font></h2>



---

#### Forward Method

---

Formula:

f'(x) = (f(x + h) - f(x)) / h



In [124]:
# Given Function

# f(x) = 3x^2 + 2

def f(x):
    return 3 * x**2 + 2

# Forward difference method

def forwardDifference(x, h=1e-6):
    return (f(x + h) - f(x)) / h

# Compute forward difference method

x = np.arange(1,5)

print(f"\nForward difference method at x = {x}:")
print(forwardDifference(x))

# Now Checking the derivative using analytical method

def derivative(x): # Returning the derivative of function
    return 6 * x

print(f"\nAnalytical derivative at x = {x}:")
print(derivative(x))





Forward difference method at x = [1 2 3 4]:
[ 6.000003   12.000003   18.000003   24.00000299]

Analytical derivative at x = [1 2 3 4]:
[ 6 12 18 24]


---
#### Backword Method

---

Formula:

f'(x) = (f(x) - f(x - h)) / h

In [126]:
# Given Function

# f(x) = 3x^2 + 2

def f(x):
    return 3 * x**2 + 2

# Backward difference method

def backwardDifference(x,h = 1e-6):
    return ((f(x) - f(x-h))/h)

# Compute backward difference method

x = np.arange(1,5)

print(f"\nBackward difference method at x = {x}:")
print(backwardDifference(x))



Backward difference method at x = [1 2 3 4]:
[ 5.999997 11.999997 17.999997 23.999997]


---
#### Central difference method

---

Formula:

f'(x) = (f(x + h) - f(x - h)) / (2h)


In [128]:

def f(x):
    return 3 * x**2 + 2

# Central difference method

def centralDifference(x,h = 1e-6):
    return ((f(x+h) - f(x-h))/(2*h))

# Compute central difference method

x = np.arange(1,5)

print(f"\nCentral difference method at x = {x}:")
print(centralDifference(x))


Central difference method at x = [1 2 3 4]:
[ 6. 12. 18. 24.]



---
<h2 align = center> <font color = '#7F38EC'> 

 2.Numerical Integration


  </font></h2>

---



#### What is Numerical Integration

Numerical integration is a technique used to approximate the definite integral of a function by approximating the area under the curve of the function over a given interval. There are two main types of numerical integration methods: trapezoidal rule and Simpson's rule.

---

<h2 align = center> <font color = 'Silver'> 

Use NumPy to compute the numerical integral of a given function.

  </font></h2>


  ---



In [134]:

# Define the function
# f(x) = x^2 + 3x + 2
def f(x):
    return x**2 + 3*x + 2

# Intervals [a, b]
a, b = 0, 4

# Number of points
num_points = 500

# Create an array of x values from a to b
x = np.linspace(a, b, num_points)


# Compute the numerical integral using the trapezoidal rule
integral = np.trapezoid(f(x), x)

print("Numerical integral of f(x) from {} to {} is: {}".format(a, b, integral))


Numerical integral of f(x) from 0 to 4 is: 53.333376171180035


---
<h2 align = center> <font color = 'Cheese'> 
Implement the trapezoidal rule and Simpson's rule for integration.
  </font></h2>

---


---
### Using Trapezoidal Rule

---


Trapezoidal Rule:

Integral[a, b] = (h/2) * [f(x0) + 2*f(x1) + 2*f(x2) +... + 2*f(xn-1) + f(xn)]

In [136]:


# Define the function
# f(x) = x^2 + 3x + 2
def f(x):
    return x**2 + 3*x + 2

# Intervals [a, b]
a, b = 0, 4

# Number of points
num_points = 500

# Create an array of x values from a to b
x = np.linspace(a, b, num_points)


# Compute the numerical integral using the trapezoidal rule
integral = np.trapezoid(f(x), x)

print(f"Numerical integral of f(x) from {a} to {b} is: {integral}")



Numerical integral of f(x) from 0 to 4 is: 53.333376171180035


---  
### Using Simson's Rule For Integration

---

Simpson's 1/3 Rule:

Integral[a, b] = (h/3) * [f(x0) + 4*f(x1) + 2*f(x2) +... + 4*f(xn-2) + f(xn-1)]


where:

h = (b - a) / (num_points - 1)


In [141]:
import numpy as np

# Define the function
def f(x):
    return x**2 + 3*x + 2

# Intervals [a, b]
a, b = 0, 4

# Number of points
num_points = 500

# Create an array of x values from a to b
x = np.linspace(a, b, num_points)


# Compute the numerical integral using Simpson's rule
h = (b - a) / (num_points - 1)
integral = f(x[0]) + f(x[-1]) + 4 * np.sum(f(x[1:-1:2])) + 2 * np.sum(f(x[2:-2:2]))
integral *= h / 3

print(f"Numerical integral of f(x) from {a} to {b} is: {integral}")


Numerical integral of f(x) from 0 to 4 is: 53.093441048412075



---
<h2 align = center> <font color = '#6F2DA8'>3. Partial Derivatives</font></h2>

---




---

<h2 align = center> <font color = 'crimson'> 
Calculate partial derivatives of multivariable functions using NumPy.
  </font></h2>


  ---






#### What is partial derivatives ?

Partial derivatives are used to find the rate of change of a function with respect to one of its variables while keeping the others constant. They are represented as partial derivatives with respect to each variable.



### Formula

<img src="formula.png" alt="Partial Derivative">






In [161]:
# Define a function

def f(x, y):
    return x**2 + 3*x*y + 2*y**2

# Compute partial derivatives
x = 2
y= 5

dx = 1e-6
dy = 1e-6


# Compute partial derivative with respect to x

partial_x = (f(x + dx, y)-f(x,y))/dx

# Compute partial derivative with respect to y

partial_y = (f(x,y+dy)-f(x,y))/dy

# Print the partial derivatives

print(f"Partial derivative of f({x}, {y}) with respect to x: {partial_x}")
print(f"Partial derivative of f({x}, {y}) with respect to y: {partial_y}")




Partial derivative of f(2, 5) with respect to x: 19.00000100363286
Partial derivative of f(2, 5) with respect to y: 26.0000020091411




<h2 align = center> <font color = 'neonGreen'> 
Verify results by comparing with analytical solutions.

  </font></h2>








In [163]:

#### Analytical solutions

# Analytical solutions to the partial derivatives of the given function f(x, y) are:

# Function: x**2 + 3*x*y + 2*y**2

# Compute the analytical solutions

analytical_partial_x = 2*x + 3*y # This equation is accuired by taking derivative of given function by x 
analytical_partial_y = 3*x + 4*y # This equation is accuired by taking derivative of given function by y

# Print the analytical solutions

print(f"Analytical partial derivative of f({x}, {y}) with respect to x: {analytical_partial_x}")
print(f"Analytical partial derivative of f({x}, {y}) with respect to y: {analytical_partial_y}")


# ---

Analytical partial derivative of f(2, 5) with respect to x: 19
Analytical partial derivative of f(2, 5) with respect to y: 26


### Result

Partial derivative of f(2, 5) with respect to x: 19.00000100363286
Partial derivative of f(2, 5) with respect to y: 26.0000020091411

Analytical partial derivative of f(2, 5) with respect to x: 19   
Analytical partial derivative of f(2, 5) with respect to y: 26


As we can see , the partial derivatives solution find by numeric methods and analytical methods are approximately same.


---
<h2 align = center> <font color = '	#228B22'> 4. Optimization</font></h2>

---



#### What is optimization?

Optimization is the process of finding the best possible outcome or values for a set of variables, given a specific objective function and constraints. It can be used to find the best solutions to problems like finding the best parameters for a machine learning model, finding the optimal location for a production facility, or finding the best allocation of resources to meet demand.

Optimization problems can be classified into two main types: unconstrained and constrained optimization problems. Unconstrained optimization problems have no constraints on the variables, while constrained optimization problems have some restrictions or constraints on the variables.


---


<h2 align = center> <font color = '#FFFF00'> Use NumPy to solve optimization problems with constraints.</font></h2>

---








#### What are optimization problems?

Optimization problems involve finding the best possible solution to a specific objective function subject to certain constraints. Examples of optimization problems include finding the minimum or maximum value of a function, finding the best combination of variables to achieve a desired outcome, or finding the best allocation of resources to meet a given deadline.

Optimization problems can be solved using various algorithms, such as gradient descent, gradient ascent, and simulated annealing, to find the optimal solution.

---
### Maximization Problem

---

Problem Statement:
A company produces three products: Product X, Product Y, and Product Z. The company aims to maximize its profit from these products. The profit margins per unit sold are as follows:

- Product X yields a profit of $3 per unit.
- Product Y yields a profit of $5 per unit.
- Product Z yields a profit of $4 per unit.


However, the company faces constraints in production and resources:

1. Each unit of Product X requires 2 units of a specific raw material and 3 units of labor, and the total availability of these resources is limited to 8 units.

2. Each unit of Product Y requires 2 units of labor and 5 units of another raw material, with a total of 10 units available for these resources.

3. Each unit of Product Z requires 3 units of Product X, 2 units of Product Y, and 4 units of another raw material, with a total of 15 units available for these resources.

4. The company cannot produce a negative number of any product.


The objective is to determine how many units of each product the company should produce to maximize its total profit while adhering to these constraints.

### Objective:

Maximize: Z = 3x + 5y + 4z


### Equation


Constraints:

2x + 3y <= 8   
2y + 5z <= 10  
3x + 2y + 4z <= 15   
x >= 0 and y >= 0 and z >= 0

In [213]:
from scipy.optimize import linprog

# Cost Matrix
cost = [-3, -5, -4]

# Left-hand side of equations
leftHandSide = [
    [2, 3, 0],  
    [0, 2, 5],  
    [3, 2, 4]   
]

# Right-hand side of equations
rightHandSide = [8, 10, 15]

# Solving Maximization Problem
optimumSolution = linprog(c=cost, A_ub=leftHandSide, b_ub=rightHandSide, method="highs")

# Printing Result
print(optimumSolution)


        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -18.65853658536585
              x: [ 2.171e+00  1.220e+00  1.512e+00]
            nit: 3
          lower:  residual: [ 2.171e+00  1.220e+00  1.512e+00]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  0.000e+00  0.000e+00]
                 marginals: [-1.098e+00 -5.854e-01 -2.683e-01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0


### Interpretation

The solution provided by the solver suggests that producing approximately 
2.171
2.171 units of Product X, 
1.22
1.22 units of Product Y, and 
1.512
1.512 units of Product Z will maximize the profit, Which will be 18.65 USD










---
### Minimization Problem

---

Problem Statment:

A manufacturing company is producing two products, Product A and Product B. The company aims to minimize the total production cost. The cost of producing one unit of Product A is $1, and the cost of producing one unit of Product B is also $1.

However, the production process is subject to the following constraints:

Resource Constraint 1:

The production of Product A and Product B requires certain raw materials. Specifically, 2 units of raw material are needed for each unit of Product A, and 4 units of raw material are needed for each unit of Product B. The total amount of this raw material available must be at least 4 units.
Resource Constraint 2:

There is another resource constraint where each unit of Product A requires 1 unit of a secondary material, and each unit of Product B requires 7 units of this secondary material. The total amount of this secondary material available must be at least 7 units.
Non-negativity Constraint:

The company cannot produce a negative number of units of either product, meaning the quantities of both products must be non-negative.
The company needs to determine the number of units of Product A (x1) and Product B (x2) to produce in order to minimize the total production cost while satisfying these constraints.

### Objective

Minimize: Z = x1 + x2

### Equation

Constraints:

2x1 + 4x2 >= 4  --> (1)   
x1 + 7x2 >= 7  --> (2)     
x1,x2 >= 0



### Note :

Since there is greater than sign in both equations,this is easily remedied by converting the “greater than” inequality constraint to a “less than” inequality constraint by multiplying both sides by -1 .

So the both equations will change into below : 

-2x1 -4x2 <= -4 -->(1)   
-x1 -7x2 <= -7 -->(2)

In [212]:
from scipy.optimize import linprog

# Cost Matrix
costMatrix = [1, 1]

# Left-hand side of equations
leftHandSide = [
    [-2, -4], 
    [-1,-7 ]    
]

# Right-hand side of equations
rightHandSide = [-4, -7]

# Solving minimization problem
optimumSolution = linprog(c=costMatrix, A_ub=leftHandSide, b_ub=rightHandSide,  method="highs")

# Printing Results
print(optimumSolution)



        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1.0
              x: [ 0.000e+00  1.000e+00]
            nit: 0
          lower:  residual: [ 0.000e+00  1.000e+00]
                 marginals: [ 8.571e-01  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -1.429e-01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0


### Interpretation
The solution shows the minimum production cost of $1 can be achieved by Making 0 units of Product A and 1 unit of Product B while satisfying all constraints. 
